# HealthConnect Clinic - Week 4 Data Analytics

## Project Overview
The HealthConnect project focuses on understanding appointment attendance and identifying factors associated with missed appointments. The overall objective is to use data analytics to understand appointment patterns, identify factors that may contribute to missed appointments, and provide an analytical foundation for improving appointment attendance, and supporting better patient management.

## Data Overview
The HealthConnect dataset contains:
•	5,000 records.
•	18 variables.
The dataset contains categorical, numerical and date-related variables.

## Import Libraries

In [1]:
import pandas as pd 
import numpy as np

## Load Dataset

In [2]:
df = pd.read_csv("HealthConnect_Appointment_Data.csv")

## Data Inspection

In [3]:
df.head()

,appointment_id,patient_id,gender,age,age_group,appointment_type,booking_date,appointment_date,appointment_day,appointment_time,booking_lead_days,previous_appointments,previous_no_shows,reminder_sent,reminder_channel,distance_to_clinic_km,waiting_time_minutes,appointment_outcome
0,HC-00001,P-1613,Female,39,35-44,Follow-up,2/6/2025,2/18/2025,Tuesday,Afternoon,12,2,0,Yes,WhatsApp,19.3,29.0,No-Show
1,HC-00002,P-0813,Male,31,25-34,Specialist Consultation,2/25/2026,2/27/2026,Friday,Morning,2,6,0,Yes,SMS,14.3,42.0,Attended
2,HC-00003,P-1366,Female,50,45-54,General Consultation,11/16/2025,12/24/2025,Wednesday,Morning,38,5,1,Yes,SMS,11.4,11.0,No-Show
3,HC-00004,P-1031,Male,59,55-64,Follow-up,7/18/2025,8/28/2025,Thursday,Evening,41,3,1,Yes,SMS,7.4,35.0,Attended
4,HC-00005,P-1458,Female,34,25-34,Follow-up,7/9/2025,8/25/2025,Monday,Afternoon,47,3,1,Yes,Email,5.6,27.0,No-Show


In [4]:
df.tail()

,appointment_id,patient_id,gender,age,age_group,appointment_type,booking_date,appointment_date,appointment_day,appointment_time,booking_lead_days,previous_appointments,previous_no_shows,reminder_sent,reminder_channel,distance_to_clinic_km,waiting_time_minutes,appointment_outcome
4995,HC-04996,P-0007,Male,49,45-54,Diagnostic Test,3/29/2025,4/26/2025,Saturday,Morning,28,5,1,Yes,Email,4.4,35.0,No-Show
4996,HC-04997,P-1501,Female,52,45-54,Follow-up,12/14/2025,2/12/2026,Thursday,Evening,60,6,2,Yes,WhatsApp,12.6,24.0,Attended
4997,HC-04998,P-0315,Male,59,55-64,General Consultation,3/3/2026,3/24/2026,Tuesday,Afternoon,21,5,0,No,NaN,22.8,28.0,Attended
4998,HC-04999,P-1262,Male,47,45-54,General Consultation,12/3/2025,1/6/2026,Tuesday,Morning,34,4,1,No,NaN,12.5,16.0,Attended
4999,HC-05000,P-1793,Female,19,18-24,General Consultation,8/24/2025,9/30/2025,Tuesday,Morning,37,6,1,Yes,WhatsApp,1.1,13.0,No-Show


In [5]:
df.shape

(5000, 18)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   appointment_id         5000 non-null   object 
 1   patient_id             5000 non-null   object 
 2   gender                 5000 non-null   object 
 3   age                    5000 non-null   int64  
 4   age_group              5000 non-null   object 
 5   appointment_type       5000 non-null   object 
 6   booking_date           5000 non-null   object 
 7   appointment_date       5000 non-null   object 
 8   appointment_day        5000 non-null   object 
 9   appointment_time       5000 non-null   object 
 10  booking_lead_days      5000 non-null   int64  
 11  previous_appointments  5000 non-null   int64  
 12  previous_no_shows      5000 non-null   int64  
 13  reminder_sent          5000 non-null   object 
 14  reminder_channel       3634 non-null   object 
 15  dist

### Convert Date Columns to Datetime
The booking and appointment date columns are converted from object datatype into datetime datatype. This is necessary because date calculations, comparisons, and time-based analysis require proper datetime types.

In [7]:
df["booking_date"] = pd.to_datetime(df["booking_date"])
df["appointment_date"] = pd.to_datetime(df["appointment_date"])

In [8]:
df.dtypes

appointment_id                   object
patient_id                       object
gender                           object
age                               int64
age_group                        object
appointment_type                 object
booking_date             datetime64[ns]
appointment_date         datetime64[ns]
appointment_day                  object
appointment_time                 object
booking_lead_days                 int64
previous_appointments             int64
previous_no_shows                 int64
reminder_sent                    object
reminder_channel                 object
distance_to_clinic_km           float64
waiting_time_minutes            float64
appointment_outcome              object
dtype: object

In [9]:
df.isnull().sum().sort_values(ascending=False)

reminder_channel         1366
distance_to_clinic_km      90
waiting_time_minutes       60
appointment_id              0
patient_id                  0
reminder_sent               0
previous_no_shows           0
previous_appointments       0
booking_lead_days           0
appointment_time            0
appointment_day             0
appointment_date            0
booking_date                0
appointment_type            0
age_group                   0
age                         0
gender                      0
appointment_outcome         0
dtype: int64

In [10]:
null_percentage = (df.isnull().sum().sort_values(ascending=False) / len(df)) * 100
print(null_percentage)

reminder_channel         27.32
distance_to_clinic_km     1.80
waiting_time_minutes      1.20
appointment_id            0.00
patient_id                0.00
reminder_sent             0.00
previous_no_shows         0.00
previous_appointments     0.00
booking_lead_days         0.00
appointment_time          0.00
appointment_day           0.00
appointment_date          0.00
booking_date              0.00
appointment_type          0.00
age_group                 0.00
age                       0.00
gender                    0.00
appointment_outcome       0.00
dtype: float64


In [11]:
df.duplicated().sum()

np.int64(0)

### Check for Duplicate Appointment IDs
Appointment IDs should uniquely identify individual appointments. This check confirms whether the identifier is repeated across records.

In [12]:
df["appointment_id"].duplicated().sum()

np.int64(0)

### Examine Categorical Value Distributions
This identifies object columns and prints the frequency of every category, including missing values. This helps understand the composition of categorical variables and identify unexpected or unusual values.

In [13]:
for col in df.select_dtypes(include="object").columns:
    print(f"\n{col}")
    print(df[col].value_counts(dropna=False))


appointment_id
appointment_id
HC-00001    1
HC-03331    1
HC-03338    1
HC-03337    1
HC-03336    1
           ..
HC-01667    1
HC-01666    1
HC-01665    1
HC-01664    1
HC-05000    1
Name: count, Length: 5000, dtype: int64

patient_id
patient_id
P-1399    9
P-1742    9
P-1776    9
P-0887    8
P-0252    8
         ..
P-1038    1
P-0255    1
P-1043    1
P-1650    1
P-1262    1
Name: count, Length: 1696, dtype: int64

gender
gender
Female               2488
Male                 2404
Prefer not to say     108
Name: count, dtype: int64

age_group
age_group
65+      1241
35-44     819
55-64     800
45-54     793
25-34     783
18-24     564
Name: count, dtype: int64

appointment_type
appointment_type
General Consultation       2086
Follow-up                  1421
Specialist Consultation     900
Diagnostic Test             593
Name: count, dtype: int64

appointment_day
appointment_day
Wednesday    737
Sunday       737
Friday       728
Saturday     724
Monday       701
Tuesday      689
Thursd

No obvious spelling variations, capitalization problems, unexpected categories, or duplicate category representations were identified. Therefore, no categorical standardisation or recoding is required.

### Descriptive Statistics
These statistics provide an initial understanding of the distribution and range of the numerical variables.

In [14]:
df.describe()

,age,booking_date,appointment_date,booking_lead_days,previous_appointments,previous_no_shows,distance_to_clinic_km,waiting_time_minutes
count,5000.000000,5000,5000,5000.00000,5000.000000,5000.000000,4910.000000,4940.000000
mean,48.794800,2025-08-31 03:01:09.120000,2025-09-29 18:20:44.160000,29.63860,3.013800,0.544200,10.109572,24.189676
min,18.000000,2024-11-07 00:00:00,2025-01-01 00:00:00,0.00000,0.000000,0.000000,0.500000,2.000000
25%,33.000000,2025-04-15 00:00:00,2025-05-15 00:00:00,15.00000,2.000000,0.000000,5.300000,17.000000
50%,49.000000,2025-08-28 00:00:00,2025-09-27 00:00:00,30.00000,3.000000,0.000000,8.700000,24.000000
75%,64.000000,2026-01-16 06:00:00,2026-02-15 00:00:00,45.00000,4.000000,1.000000,13.500000,32.000000
max,80.000000,2026-06-27 00:00:00,2026-06-30 00:00:00,60.00000,11.000000,5.000000,45.000000,68.000000
std,18.138547,NaN,NaN,17.39936,1.741211,0.746832,6.590030,10.863184


### Validate Previous No-Show History
A patient's number of previous no-shows should not exceed their number of previous appointments. This counts records that violate that logical relationship.

In [15]:
(df["previous_no_shows"] > df["previous_appointments"]).sum()

np.int64(0)

This confirms that there are no records where the number of previous no-shows exceeds the total number of previous appointments

### Validate Appointment and Booking Dates
An appointment should normally occur after it is booked. This counts records where the appointment date is later than the booking date.

In [16]:
(df["appointment_date"] > df["booking_date"]).sum()

np.int64(4915)

4,915 appointments occurred after booking

### Validate the Booking Lead-Time Field
The actual lead time is independently calculated by subtracting the booking date from the appointment date. It is then compared with the existing "booking_lead_days" column to confirm that the recorded values are accurate.

In [17]:
calculated_lead_days = (df["appointment_date"] - df["booking_date"]).dt.days
(df["booking_lead_days"] != calculated_lead_days).sum()

np.int64(0)

This means there were zero discrepancies across all 5,000 records. Therefore, the booking_lead_days variable is completely consistent with the underlying booking and appointment dates.

### Check for Impossible Date Order
This identifies appointments whose appointment date occurs before the booking date.

In [18]:
(df["appointment_date"] < df["booking_date"]).sum()

np.int64(0)

0 appointments occurred before booking

### Check Same-Day Bookings
This counts appointments where the booking date and appointment date are exactly the same.

In [19]:
(df["appointment_date"] == df["booking_date"]).sum()

np.int64(85)

85 appointments were booked and scheduled on the same day

### Recheck Missing Values Before Analysis
The missing-value summary is repeated after the initial data-quality checks. This confirms which columns still contain missing observations before deeper analysis begins.

In [20]:
df.isnull().sum().sort_values(ascending=False)

reminder_channel         1366
distance_to_clinic_km      90
waiting_time_minutes       60
appointment_id              0
patient_id                  0
reminder_sent               0
previous_no_shows           0
previous_appointments       0
booking_lead_days           0
appointment_time            0
appointment_day             0
appointment_date            0
booking_date                0
appointment_type            0
age_group                   0
age                         0
gender                      0
appointment_outcome         0
dtype: int64

### Examine Reminder-Channel Missingness
A cross-tabulation compares whether a reminder was sent with the reminder channel used. This helps verify whether missing reminder channels are logically associated with appointments where no reminder was sent

In [21]:
pd.crosstab(
    df["reminder_sent"],
    df["reminder_channel"],
    dropna=False
)

reminder_channel,Email,SMS,WhatsApp,NaN
reminder_sent,,,,
No,0,0,0,1366
Yes,533,2000,1101,0


All 1,366 records where no reminder was sent had a missing reminder channel. Conversely, every appointment where a reminder was sent had a recorded channel.
The missing values were classified as structural missingness rather than data-entry errors. They are therefore retained as NaN.

### Inspect Records with Missing Distance Values
This filters the dataset to appointments where the distance to the clinic is missing. Viewing actual records helps determine whether the missing values appear random or follow a recognizable pattern.

In [22]:
df[df["distance_to_clinic_km"].isnull()].head()

,appointment_id,patient_id,gender,age,age_group,appointment_type,booking_date,appointment_date,appointment_day,appointment_time,booking_lead_days,previous_appointments,previous_no_shows,reminder_sent,reminder_channel,distance_to_clinic_km,waiting_time_minutes,appointment_outcome
38,HC-00039,P-0072,Male,78,65+,General Consultation,2025-04-02,2025-06-01,Sunday,Afternoon,60,6,1,Yes,SMS,NaN,42.0,Cancelled
176,HC-00177,P-1036,Male,65,65+,General Consultation,2026-01-15,2026-01-27,Tuesday,Morning,12,2,0,Yes,Email,NaN,28.0,No-Show
443,HC-00444,P-1114,Female,53,45-54,Follow-up,2025-08-06,2025-09-24,Wednesday,Afternoon,49,4,1,No,NaN,NaN,18.0,No-Show
503,HC-00504,P-0705,Female,21,18-24,General Consultation,2026-03-06,2026-04-19,Sunday,Afternoon,44,3,0,Yes,SMS,NaN,36.0,No-Show
581,HC-00582,P-1596,Male,38,35-44,Follow-up,2026-02-10,2026-04-01,Wednesday,Morning,50,1,0,Yes,SMS,NaN,11.0,No-Show


This inspection shows that the records still contained information for the other variables

### Compare Distance Missingness with Appointment Outcomes
This calculates the percentage distribution of appointment outcomes for records with and without distance information. The purpose is to determine whether missing distance values are associated with noticeably different outcomes.

In [23]:
pd.crosstab(
    df["distance_to_clinic_km"].isnull(),
            df["appointment_outcome"],
            normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
distance_to_clinic_km,,,
False,46.334012,5.274949,48.391039
True,43.333333,4.444444,52.222222


The no-show rate differed by approximately 3.83% between the two groups. The outcome distributions are considered broadly similar, and there was no strong evidence from this assessment that the missing values are heavily associated with a particular outcome.
The 90 missing values are retained as NaN.

### Inspect Records with Missing Waiting-Time Values
This displays appointments where waiting time is missing. Examining these records helps understand the context of the missing observations.

In [24]:
df[df["waiting_time_minutes"].isnull()].head()

,appointment_id,patient_id,gender,age,age_group,appointment_type,booking_date,appointment_date,appointment_day,appointment_time,booking_lead_days,previous_appointments,previous_no_shows,reminder_sent,reminder_channel,distance_to_clinic_km,waiting_time_minutes,appointment_outcome
54,HC-00055,P-1767,Female,45,45-54,Follow-up,2025-09-26,2025-09-29,Monday,Evening,3,2,0,No,NaN,11.3,NaN,Attended
161,HC-00162,P-0306,Female,28,25-34,Diagnostic Test,2025-12-28,2025-12-29,Monday,Morning,1,3,1,Yes,SMS,17.8,NaN,No-Show
191,HC-00192,P-0482,Female,53,45-54,Follow-up,2025-10-17,2025-12-11,Thursday,Afternoon,55,4,2,Yes,WhatsApp,9.7,NaN,No-Show
260,HC-00261,P-1569,Male,29,25-34,General Consultation,2026-01-09,2026-03-10,Tuesday,Afternoon,60,5,1,Yes,SMS,9.8,NaN,No-Show
343,HC-00344,P-0344,Male,57,55-64,Follow-up,2025-09-12,2025-10-09,Thursday,Morning,27,2,0,Yes,WhatsApp,4.6,NaN,Attended


### Compare Waiting-Time Missingness with Appointment Outcomes
This compares appointment outcomes for records where waiting time is missing versus available. The percentages help assess whether missing waiting-time data may be related to appointment outcome.

In [25]:
pd.crosstab(
    df["waiting_time_minutes"].isnull(),
            df["appointment_outcome"],
            normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
waiting_time_minutes,,,
False,46.417004,5.283401,48.299595
True,35.000000,3.333333,61.666667


### Examine Waiting-Time Missingness Within Each Outcome
This reverses the perspective by calculating the percentage of missing waiting-time values within each appointment outcome. It helps determine whether missingness is concentrated in a particular outcome category.


In [26]:
pd.crosstab(
    df["appointment_outcome"],
            df["waiting_time_minutes"].isnull(),
            normalize="index"
) * 100

waiting_time_minutes,False,True
appointment_outcome,,
Attended,99.092481,0.907519
Cancelled,99.239544,0.760456
No-Show,98.472967,1.527033


Although the no-show group had the highest proportion of missing waiting-time values, the missingness remained low across all outcome categories.
The 60 missing values are retained as NaN.
For analyses specifically involving waiting time, the available observations will be used and the missing observations will be acknowledged as a limitation.

## Exploratory Data Analysis

### Count Appointment Outcomes
This counts the number of appointments in each outcome category. It provides the basic distribution of attended, no-show, and cancelled appointments.

In [27]:
df["appointment_outcome"].value_counts()

appointment_outcome
No-Show      2423
Attended     2314
Cancelled     263
Name: count, dtype: int64

### Calculate Appointment Outcome Percentages
This makes it easier to compare the relative size of each outcome category.

In [28]:
df["appointment_outcome"].value_counts(normalize=True) * 100

appointment_outcome
No-Show      48.46
Attended     46.28
Cancelled     5.26
Name: proportion, dtype: float64

This shows that No-Show is the most common appointment outcome, accounting for 2,423 appointments (48.46%), Attended appointments account for 2,314 appointments (46.28%), while 263 appointments (5.26%) were cancelled.
This means that almost one out of every two scheduled appointments resulted in a no-show. The relatively high proportion of missed appointments establishes No-Shows as an important issue for further investigation.

### Analyse Outcomes by Appointment Type
The data is grouped by appointment type and then the outcome categories are counted. This shows how attended, no-show, and cancelled appointments are distributed within each type of appointment.

In [29]:
df.groupby("appointment_type")["appointment_outcome"].value_counts()

appointment_type         appointment_outcome
Diagnostic Test          No-Show                 295
                         Attended                272
                         Cancelled                26
Follow-up                No-Show                 728
                         Attended                616
                         Cancelled                77
General Consultation     Attended               1007
                         No-Show                 973
                         Cancelled               106
Specialist Consultation  No-Show                 427
                         Attended                419
                         Cancelled                54
Name: count, dtype: int64

This shows that General Consultation had the highest number of appointments overall, followed by Follow-up, Specialist Consultation, and Diagnostic Test.

### Calculate Outcome Rates by Appointment Type
This cross-tabulation calculates the percentage of each outcome within every appointment type. It allows fair comparison between appointment types even when their total volumes differ.

In [30]:
pd.crosstab(
    df["appointment_type"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
appointment_type,,,
Diagnostic Test,45.868465,4.384486,49.747049
Follow-up,43.349754,5.418719,51.231527
General Consultation,48.274209,5.081496,46.644295
Specialist Consultation,46.555556,6.000000,47.444444


The highest no-show rate was observed among Follow-up appointments at 51.23%. Diagnostic Tests had the second-highest no-show rate at 49.75%, followed by Specialist Consultations at 47.44%. General Consultations had the lowest no-show rate at 46.64%.
This indicates that appointment type may be associated with differences in attendance behaviour, although the differences are not extremely large.
An important observation is that Follow-up appointments have a particularly high no-show rate despite General Consultations having the largest number of appointments. This means that the result is not simply being driven by appointment volume.

### Compare Reminders with Appointment Outcomes
This cross-tabulation shows the number of attended, no-show, and cancelled appointments for records that received or did not receive reminders.

In [31]:
pd.crosstab(
    df["reminder_sent"],
    df["appointment_outcome"],
)

appointment_outcome,Attended,Cancelled,No-Show
reminder_sent,,,
No,583,81,702
Yes,1731,182,1721


### Calculate Outcome Rates by Reminder Status
The cross-tabulation is normalised by row to show the percentage of each outcome among appointments with and without reminders. This helps assess whether reminder status is associated with attendance behaviour.

In [32]:
pd.crosstab(
    df["reminder_sent"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
reminder_sent,,,
No,42.679356,5.929722,51.390922
Yes,47.633462,5.008255,47.358283


Among the 1,366 appointments without a reminder, 51.39% resulted in a no-show and 42.68% were attended. Among the 3,634 appointments where a reminder was sent, 47.36% resulted in a no-show and 47.63% were attended. The no-show rate was therefore 4.03 percentage points lower among appointments where a reminder was sent, while the attendance rate was approximately 4.95 percentage points higher. 
These results indicate an observed association between reminder status and appointment outcomes. However, this analysis does not establish causation, and other appointment or patient characteristics may also influence the relationship.

### Count Reminder Channels Used
Only appointments where a reminder was actually sent are selected. The code then counts how many reminders were delivered through SMS, WhatsApp, and Email.

In [33]:
reminder_channel_counts = df[df["reminder_sent"] == "Yes"]["reminder_channel"].value_counts()
print(reminder_channel_counts)

reminder_channel
SMS         2000
WhatsApp    1101
Email        533
Name: count, dtype: int64


### Compare Reminder Channels with Outcomes
This cross-tabulation shows the number of attended, no-show, and cancelled appointments for each reminder channel. It provides the volume needed to compare reminder-channel performance.

In [34]:
pd.crosstab(
    df[df["reminder_sent"] == "Yes"]["reminder_channel"],
    df[df["reminder_sent"] == "Yes"]["appointment_outcome"]
) 

appointment_outcome,Attended,Cancelled,No-Show
reminder_channel,,,
Email,248,27,258
SMS,992,93,915
WhatsApp,491,62,548


### Calculate Outcome Rates by Reminder Channel
The results are normalised by reminder channel so that each channel can be compared using percentages rather than raw counts.

In [35]:
pd.crosstab(
    df[df["reminder_sent"] == "Yes"]["reminder_channel"],
    df[df["reminder_sent"] == "Yes"]["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
reminder_channel,,,
Email,46.529081,5.065666,48.405253
SMS,49.600000,4.650000,45.750000
WhatsApp,44.595822,5.631244,49.772934


Only appointments where a reminder was sent were included, excluding the 1,366 records with no reminder and therefore no reminder channel. SMS is the most frequently used channel, accounting for 2,000 appointments, followed by WhatsApp with 1,101 and Email with 533. When outcome percentages were compared within each channel, SMS had the lowest observed no-show rate at 45.75%, followed by Email at 48.41%, while WhatsApp had the highest at 49.77%. SMS also recorded the highest attendance rate at 49.60%. 
The 4.02-percentage-point difference between the highest and lowest no-show rates suggests that reminder channel may be associated with appointment outcomes. However, this result does not establish that one channel causes better attendance, as other factors may influence the observed differences.

### Analyse Booking Lead Time by Appointment Outcome
This calculates the count, mean, median, minimum, and maximum booking lead days for each appointment outcome. It helps determine whether how far in advance an appointment is booked is related to attendance behaviour.

In [36]:
df.groupby("appointment_outcome",)["booking_lead_days"].agg(["count", "mean", "median", "min", "max"])

,count,mean,median,min,max
appointment_outcome,,,,,
Attended,2314,24.521175,23.0,0,60
Cancelled,263,29.631179,30.0,0,60
No-Show,2423,34.526620,37.0,0,60


Attended appointments had an average booking lead time of 24.52 days and a median of 23 days, while no-show appointments had an average of 34.53 days and a median of 37 days. This represents an average difference of approximately 10 days and a median difference of 14 days. Cancelled appointments fell between the two groups, with an average lead time of 29.63 days. 
This result indicate an observed association between longer booking lead times and no-shows. However, the analysis does not establish causation, and other factors may contribute to this relationship.

### Examine Previous No-Show Frequency
This counts how many patients have each number of previous no-shows. Sorting by the number of no-shows makes the distribution easier to interpret.

In [37]:
df["previous_no_shows"].value_counts().sort_index()

previous_no_shows
0    2921
1    1548
2     438
3      78
4      12
5       3
Name: count, dtype: int64

### Analyse Previous No-shows Against Current Outcomes
This calculates the percentage of appointment outcomes for each level of previous no-shows. It helps investigate whether previous no-show behaviour is associated with future attendance.

In [38]:
pd.crosstab(
    df["previous_no_shows"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
previous_no_shows,,,
0,50.462170,6.025334,43.512496
1,42.248062,4.263566,53.488372
2,36.301370,4.337900,59.360731
3,29.487179,2.564103,67.948718
4,33.333333,0.000000,66.666667
5,0.000000,0.000000,100.000000


This result showed a clear increase in current no-show rates as the number of previous no-shows increased. Patients with no previous no-shows had a current no-show rate of 43.51%, compared with 53.49.

### Examine the Age-Group Distribution
This counts appointments within each predefined age group. It provides an overview of which age groups account for the largest share of appointments.

In [39]:
df["age_group"].value_counts()

age_group
65+      1241
35-44     819
55-64     800
45-54     793
25-34     783
18-24     564
Name: count, dtype: int64

### Analyse Outcomes by Age Group
The cross-tabulation calculates the percentage of appointment outcomes within each age group. This helps identify whether attendance patterns differ across age categories.

In [40]:
pd.crosstab(
    df["age_group"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
age_group,,,
18-24,45.567376,4.255319,50.177305
25-34,45.083014,4.214559,50.702427
35-44,45.665446,5.982906,48.351648
45-54,46.027743,5.926860,48.045397
55-64,45.000000,4.250000,50.750000
65+,48.751007,6.124093,45.124899


No-show rates were relatively similar across most age groups, ranging from 48.05% to 50.75%, while the 65+ group recorded a lower no-show rate of 45.12% and the highest attendance rate of 48.75%. The highest no-show rate was observed among patients aged 55–64 at 50.75%, closely followed by the 25–34 group at 50.70%. 
Overall, this result suggest a modest association between age group and appointment outcomes, although the differences are smaller than those observed for factors such as previous no-show history and booking lead time.

### Examine Gender Distribution
This counts the number of appointments by gender category. It provides a basic demographic breakdown of the dataset.

In [41]:
df["gender"].value_counts()

gender
Female               2488
Male                 2404
Prefer not to say     108
Name: count, dtype: int64

### Compare Gender with Appointment Outcomes
This cross-tabulation shows the raw number of outcomes for each gender category. It allows comparison of attendance, no-show, and cancellation volumes.

In [42]:
pd.crosstab(
    df["gender"],
    df["appointment_outcome"]
)

appointment_outcome,Attended,Cancelled,No-Show
gender,,,
Female,1153,130,1205
Male,1103,130,1171
Prefer not to say,58,3,47


### Calculate Outcome Rates by Gender
The results are normalised by gender to compare outcome percentages rather than raw appointment counts.

In [43]:
pd.crosstab(
    df["gender"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
gender,,,
Female,46.342444,5.225080,48.432476
Male,45.881864,5.407654,48.710483
Prefer not to say,53.703704,2.777778,43.518519


Female and male patients had similar numbers of appointments, with 2,488 and 2,404 records respectively. The no-show rate was 48.43% among female patients and 48.71% among male patients, representing a difference of only 0.28 percentage points. The "Prefer not to say" group recorded a lower no-show rate of 43.52%, but contained only 108 appointments and was therefore interpreted cautiously. 
Overall, this result indicate that gender does not appear to be a strong differentiating factor in appointment outcomes within this dataset.

### Examine Appointment Volume by Day
This counts appointments scheduled for each day of the week. It helps identify which days carry the highest appointment volumes.

In [44]:
df["appointment_day"].value_counts()

appointment_day
Wednesday    737
Sunday       737
Friday       728
Saturday     724
Monday       701
Tuesday      689
Thursday     684
Name: count, dtype: int64

### Compare Appointment Days with Outcomes
This cross-tabulation shows the number of attended, no-show, and cancelled appointments for each day of the week.

In [45]:
pd.crosstab(
    df["appointment_day"],
    df["appointment_outcome"]
)

appointment_outcome,Attended,Cancelled,No-Show
appointment_day,,,
Friday,357,32,339
Monday,307,46,348
Saturday,344,41,339
Sunday,332,33,372
Thursday,309,40,335
Tuesday,338,27,324
Wednesday,327,44,366


### Calculate Outcome Rates by Appointment Day
The cross-tabulation is normalised by day so that attendance behaviour can be compared across weekdays without being affected by differences in appointment volume.

In [46]:
pd.crosstab(
    df["appointment_day"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
appointment_day,,,
Friday,49.038462,4.395604,46.565934
Monday,43.794579,6.562054,49.643367
Saturday,47.513812,5.662983,46.823204
Sunday,45.047490,4.477612,50.474898
Thursday,45.175439,5.847953,48.976608
Tuesday,49.056604,3.918723,47.024673
Wednesday,44.369064,5.970149,49.660787


Appointment volumes were relatively evenly distributed, ranging from 684 appointments on Thursday to 737 on Wednesday and Sunday. Sunday recorded the highest no-show rate at 50.47%, followed by Wednesday at 49.66% and Monday at 49.64%. Friday had the lowest no-show rate at 46.57%, while Tuesday recorded a similar rate of 47.02%. The difference between the highest and lowest no-show rates was 3.90 percentage points. 
Overall, appointment day shows some variation in attendance outcomes, but the differences are relatively modest compared with stronger patterns observed for previous no-show history and booking lead time.

### Examine Appointment-Time Distribution
This counts appointments scheduled in the morning, afternoon, and evening. It provides an overview of appointment demand across time periods.

In [47]:
df["appointment_time"].value_counts()

appointment_time
Morning      2227
Afternoon    2094
Evening       679
Name: count, dtype: int64

### Compare Appointment Time with Outcomes
This shows the raw outcome counts for morning, afternoon, and evening appointments.

In [48]:
pd.crosstab(
    df["appointment_time"],
    df["appointment_outcome"]
)

appointment_outcome,Attended,Cancelled,No-Show
appointment_time,,,
Afternoon,966,115,1013
Evening,304,37,338
Morning,1044,111,1072


### Calculate Outcome Rates by Appointment Time
The results are normalised by time period to compare attendance patterns across morning, afternoon, and evening appointments.

In [49]:
pd.crosstab(
    df["appointment_time"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
appointment_time,,,
Afternoon,46.131805,5.491882,48.376313
Evening,44.771723,5.449190,49.779087
Morning,46.879210,4.984284,48.136507


Evening appointments recorded the highest no-show rate at 49.78%, followed by Afternoon appointments at 48.38% and Morning appointments at 48.14%. The difference between the highest and lowest no-show rates was only 1.64 percentage points, indicating a relatively small variation across appointment times. 
While evening appointments may warrant further consideration, appointment time appears to have a weaker observed relationship with no-shows than variables such as booking lead time and previous no-show history.

### Examine Previous Appointment History
This counts how many appointments fall into each level of previous appointment history. It helps understand how frequently the clinic serves new versus returning patients.

In [50]:
df["previous_appointments"].value_counts().sort_index()

previous_appointments
0      242
1      749
2     1131
3     1086
4      861
5      502
6      241
7      124
8       48
9       13
10       2
11       1
Name: count, dtype: int64

### Compare Previous Appointments with Outcomes
This cross-tabulation shows the number of outcomes at each level of previous appointment history.

In [51]:
pd.crosstab(
    df["previous_appointments"],
    df["appointment_outcome"]
)

appointment_outcome,Attended,Cancelled,No-Show
previous_appointments,,,
0,124,14,104
1,362,43,344
2,534,72,525
3,483,46,557
4,407,42,412
5,219,31,252
6,105,9,127
7,59,3,62
8,14,3,31


### Calculate Outcome Rates by Previous Appointment History
The cross-tabulation is normalised by row to show the percentage of outcomes for each level of previous appointments. This helps assess whether patient familiarity with the clinic is associated with attendance.

In [52]:
pd.crosstab(
    df["previous_appointments"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
previous_appointments,,,
0,51.239669,5.785124,42.975207
1,48.331108,5.740988,45.927904
2,47.214854,6.366048,46.419098
3,44.475138,4.235727,51.289134
4,47.270616,4.878049,47.851336
5,43.625498,6.175299,50.199203
6,43.568465,3.734440,52.697095
7,47.580645,2.419355,50.000000
8,29.166667,6.250000,64.583333


No-show rates varied across the different levels of previous appointments, ranging from 42.98% among patients with no previous appointments to higher rates among some groups with more previous appointments. However, the relationship was not consistently increasing across all levels. Groups with very high numbers of previous appointments contained relatively few observations and were therefore interpreted cautiously. 
Overall, previous appointment count showed a weaker and less consistent relationship with current no-shows than previous no-show history.

### Analyse Clinic Distance by Appointment Outcome
For each appointment outcome, this calculates the number of available distance observations and summary statistics including mean, median, minimum, and maximum. It helps investigate whether distance to the clinic may be associated with attendance behaviour.

In [53]:
df.groupby("appointment_outcome")["distance_to_clinic_km"].agg(
    ["count", "mean", "median", "min", "max"]
)

,count,mean,median,min,max
appointment_outcome,,,,,
Attended,2275,9.668703,8.3,0.5,45.0
Cancelled,259,10.111583,9.1,0.6,35.4
No-Show,2376,10.531481,9.1,0.5,45.0


### Examine Distance Missingness by Outcome
This calculates the percentage of records with missing distance information within each appointment outcome. It helps assess whether missing distance data is concentrated in particular outcomes.

In [54]:
pd.crosstab(
    df["appointment_outcome"],
    df["distance_to_clinic_km"].isnull(),
    normalize="index"
) * 100

distance_to_clinic_km,False,True
appointment_outcome,,
Attended,98.314607,1.685393
Cancelled,98.479087,1.520913
No-Show,98.060256,1.939744


### Summarise Clinic-Distance Distribution
Provides the overall distribution of available distance values, including central tendency and range.

In [55]:
df["distance_to_clinic_km"].describe()

count    4910.000000
mean       10.109572
std         6.590030
min         0.500000
25%         5.300000
50%         8.700000
75%        13.500000
max        45.000000
Name: distance_to_clinic_km, dtype: float64

No-show appointments had the highest average distance at 10.53 km, compared with 9.67 km for attended appointments and 10.11 km for cancelled appointments. The difference between no-shows and attendees was approximately 0.86 km, suggesting only a modest difference. The variable contained 90 missing values, representing 1.8% of all records, with missingness distributed relatively similarly across appointment outcomes. 
Based on the descriptive analysis, distance to clinic appears to have a relatively weak observed relationship with appointment outcomes.

### Analyse Waiting Time by Appointment Outcome
For each outcome, this calculates the count, mean, median, minimum, and maximum waiting time. This supports investigation of whether waiting time differs between attended, no-show, and cancelled appointments.

In [56]:
df.groupby("appointment_outcome")["waiting_time_minutes"].agg(
    ["count", "mean", "median", "min", "max"]
)

,count,mean,median,min,max
appointment_outcome,,,,,
Attended,2293,24.289141,24.0,2.0,68.0
Cancelled,261,23.214559,22.0,2.0,55.0
No-Show,2386,24.200754,24.0,2.0,67.0


### Examine Waiting-Time Missingness by Outcome
This calculates the percentage of missing waiting-time observations within each outcome category. It helps determine whether missing waiting-time data is concentrated in specific outcomes.

In [57]:
pd.crosstab(
    df["appointment_outcome"],
    df["waiting_time_minutes"].isnull(),
    normalize="index"
) * 100

waiting_time_minutes,False,True
appointment_outcome,,
Attended,99.092481,0.907519
Cancelled,99.239544,0.760456
No-Show,98.472967,1.527033


### Summarise the Waiting-Time Distribution
Provides the overall statistical distribution of available waiting-time values.

In [58]:
df["waiting_time_minutes"].describe()

count    4940.000000
mean       24.189676
std        10.863184
min         2.000000
25%        17.000000
50%        24.000000
75%        32.000000
max        68.000000
Name: waiting_time_minutes, dtype: float64

The average waiting time was 24.29 minutes for attended appointments and 24.20 minutes for no-show appointments, representing a negligible difference of approximately 0.09 minutes. Both groups also had a median waiting time of 24 minutes. The results therefore provide little evidence of a meaningful relationship between waiting time and appointment attendance in the current dataset. The variable contained 60 missing values, representing 1.2% of the dataset, with relatively low missingness across all outcome groups.

### Create Booking Lead-Time Groups
Continuous booking lead days are converted into meaningful categories using. The groups make it easier to compare appointment outcomes across practical booking windows: 0–7, 8–14, 15–30, 31–45, and 46–60 days.

In [59]:
df["lead_time_group"] = pd.cut(
    df["booking_lead_days"],
    bins=[-1, 7, 14, 30, 45, 60],
    labels=["0-7 days", "8-14 days", "15-30 days", "31-45 days", "46-60 days"]
)

### Count Appointments by Lead-Time Group
This counts how many appointments fall within each booking lead-time category. Sorting by category order keeps the groups in their natural progression.

In [60]:
df["lead_time_group"].value_counts().sort_index()

lead_time_group
0-7 days       640
8-14 days      602
15-30 days    1333
31-45 days    1258
46-60 days    1167
Name: count, dtype: int64

### Calculate Outcomes by Lead-Time Group
The cross-tabulation calculates the percentage of each appointment outcome within every lead-time group. This helps identify whether booking further in advance is associated with higher or lower no-show rates.

In [61]:
pd.crosstab(
    df["lead_time_group"],
    df["appointment_outcome"],
    normalize="index"
) * 100

appointment_outcome,Attended,Cancelled,No-Show
lead_time_group,,,
0-7 days,66.562500,5.625000,27.812500
8-14 days,61.794020,4.651163,33.554817
15-30 days,51.687922,5.101275,43.210803
31-45 days,40.540541,5.643879,53.815580
46-60 days,27.163668,5.141388,67.694944


### Compare Raw Outcomes Across Lead-TimeGgroups
This cross-tabulation provides the actual number of attended, no-show, and cancelled appointments in each lead-time category. The raw counts complement the percentage analysis above.

In [62]:
pd.crosstab(
    df["lead_time_group"],
    df["appointment_outcome"]
)

appointment_outcome,Attended,Cancelled,No-Show
lead_time_group,,,
0-7 days,426,36,178
8-14 days,372,28,202
15-30 days,689,68,576
31-45 days,510,71,677
46-60 days,317,60,790


### Validate the New Lead-Time Grouping
This checks whether any records failed to receive a lead-time category. A zero result confirms that all booking lead-day values were successfully assigned to one of the defined groups.

In [63]:
df["lead_time_group"].isnull().sum()

np.int64(0)

Booking lead time was further analysed by grouping appointments into five intervals between 0 and 60 days. A clear and consistent increase in no-show rates was observed as booking lead time increased. Appointments booked 0–7 days in advance had a no-show rate of 27.81%, increasing to 33.55% for 8–14 days, 43.21% for 15–30 days, 53.82% for 31–45 days, and 67.69% for appointments booked 46–60 days in advance. This represents a 39.88-percentage-point difference between the shortest and longest lead-time groups. The consistent increase across every interval indicates that booking lead time is one of the strongest observed factors associated with appointment no-shows in the dataset. However, this remains an observational relationship and should not be interpreted as proof that longer booking lead times directly cause no-shows.

## Save Cleaned Dataset

In [64]:
df.to_csv("HealthConnect_Appointment_Data_Cleaned.csv", index=False)
print("cleaned dataset saved successfully")

cleaned dataset saved successfully


## Import Library

In [65]:
import os

## Confirm Existence of the Newly Saved Cleaned Dataset

In [66]:
file_path = "HealthConnect_Appointment_Data_Cleaned.csv"
if os.path.exists(file_path):
    print(f"files saved successfully:{file_path}")
else:
    print("file was not saved")

files saved successfully:HealthConnect_Appointment_Data_Cleaned.csv


In [67]:
for file in os.listdir():
    if "HealthConnect" in file:
        print(repr(file))

'HealthConnect_Appointment_Data.csv'
'HealthConnect_Appointment_Data_Cleaned.csv'


In [71]:
downloads = os.path.join(
    os.path.expanduser("~"),
    "Downloads"
)
output_file = os.path.join(
    downloads,
    "HealthConnect_Appointment_Data_Cleaned.csv"
)

df.to_csv(output_file, index=False)
print(output_file)   

C:\Users\HP\Downloads\HealthConnect_Appointment_Data_Cleaned.csv
